src/daft-plan/src/logical_optimization/rules/eliminate_cross_join.rs

```diff
@@ -0,0 +1,718 @@
+            TimeUnit::Milliseconds => true,
+            TimeUnit::Microseconds => true,
+            TimeUnit::Nanoseconds => true,
+        },
+        DataType::Utf8 => true,
+
+        DataType::Decimal128(_, _) => true,
+        DataType::Date => true,
+
+        DataType::FixedSizeBinary(_) => true,
+
+        DataType::List(_) => true,
+
+        DataType::FixedSizeList(_, _) => true,
+        DataType::Struct(fields) => fields.iter().all(|f| can_hash(&f.dtype)),
+        _ => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use daft_dsl::{col, lit};
+    use daft_schema::field::Field;
+    use rstest::*;
+
+    use super::*;
+    use crate::{
+        logical_plan::Source, source_info::PlaceHolderInfo, ClusteringSpec, LogicalPlan,
+        LogicalPlanBuilder, LogicalPlanRef, SourceInfo,
+    };
+
+    #[fixture]
+    fn t1() -> LogicalPlanRef {
+        let schema = Arc::new(
+            Schema::new(vec![
+                Field::new("a", DataType::UInt32),
+                Field::new("b", DataType::UInt32),
+                Field::new("c", DataType::UInt32),
+            ])
+            .unwrap(),
+        );
+        LogicalPlan::Source(Source {
+            output_schema: schema.clone(),
+            source_info: Arc::new(SourceInfo::PlaceHolder(PlaceHolderInfo {
+                source_schema: schema,
+                clustering_spec: Arc::new(ClusteringSpec::unknown()),
+                source_id: 0,
+            })),
+        })
+        .arced()
+    }
+
+    #[fixture]
+    fn t2() -> LogicalPlanRef {
+        let schema = Arc::new(
+            Schema::new(vec![
+                Field::new("a", DataType::UInt32),
+                Field::new("b", DataType::UInt32),
+                Field::new("c", DataType::UInt32),
+            ])
+            .unwrap(),
+        );
+        LogicalPlan::Source(Source {
+            output_schema: schema.clone(),
+            source_info: Arc::new(SourceInfo::PlaceHolder(PlaceHolderInfo {
+                source_schema: schema,
+                clustering_spec: Arc::new(ClusteringSpec::unknown()),
+                source_id: 0,
+            })),
+        })
+        .arced()
+    }
+
+    fn assert_optimized_plan_eq(plan: LogicalPlanRef, expected: Vec<&str>) {
+        let starting_schema = plan.schema();
+
+        let rule = EliminateCrossJoin::new();
+        let transformed_plan = rule.try_optimize(plan).unwrap();
+        assert!(transformed_plan.transformed, "failed to optimize plan");
+        let optimized_plan = transformed_plan.data;
+        let formatted = optimized_plan.repr_indent();
+
+        let actual: Vec<&str> = formatted.trim().lines().collect();
+        assert_eq!(
+            expected, actual,
+            "\n\nexpected:\n\n{expected:#?}\nactual:\n\n{actual:#?}\n\n"
+        );
+
+        assert_eq!(starting_schema, optimized_plan.schema())
+    }
+
+    #[rstest]
+    fn eliminate_cross_with_simple_and(t1: LogicalPlanRef, t2: LogicalPlanRef) -> DaftResult<()> {
```

Comment:
For tests, instead of checking if the printed plans are equal, could you construct an expected `LogicalPlan` and compare the optimized one to that instead? That way it won't depend on our LogicalPlan display implementation

Reply:
I think we could, But IMO, the displayed version is a bit easier to read.